### ISSUE TITLE
dummy/snippets: store service not injected, RecordArray mutated

### ISSUE THREAD

**Issue by buschtoens** [offirgolan/ember-light-table#224]
[The snippets](https://github.com/offirgolan/ember-light-table/tree/master/tests/dummy/snippets) in the dummy app assume that the `store` service is [injected](https://guides.emberjs.com/v2.8.0/applications/initializers/#toc_application-initializers) into all `Component`s just as with `Route`s, which it apparently isn't. This can confuse beginners as seen in #223.

Suggested changes:
#### Manually inject the store at the top of every snippet.

``` js
export default Ember.Component.extend({
  store: Ember.inject.service(),

  page: 1,
  // ...
});
```
#### Use `this.get('store')` instead of `this.store`.

``` js
  fetchRecords() {
    this.set('isLoading', true);
    this.get('store').query('user', this.getProperties(['page', 'limit', 'sort', 'dir'])).then(records => {
      this.get('model').pushObjects(records.toArray());
      this.set('isLoading', false);
      this.set('canLoadMore', !isEmpty(records));
    });
  },
```

---

Addendum: The original `RecordArray` is immutable. Currently we use `pushObjects` on the original `RecordArray` returned from the `model` hook. This shouldn't work, and breaks users apps in the wild.

Suggested changes:
#### Use a private `_model` that is a `model.toArray()` clone.

``` js
model: null,
_model: Ember.computed('model', function() {
  return this.get('model').toArray();
})
```

_...or..._
#### Don't use `enableSync: true` in all examples, but a special one

---

I'm down for submitting a PR.

**Comment by offirgolan** [comment]
@buschtoens good catch. PR is totally welcomed!

**Comment by trumb1mj** [comment]
When using this in a component I get `Uncaught TypeError: internalModel.getRecord is not a function(…)`

Anyone seeing this?

**Comment by trumb1mj** [comment]
How I ended up solving this was using the a private attribute like so:

```
  fetchRecords() {
    this.set('isLoading', true);
    this.get("store").query('account', {per_page: this.get("limit"), page: this.get("page")}).then(records => {
      var model = this.get('model');
      records.forEach(function(record){
        model.addObject(record._internalModel);
      });
      this.set('isLoading', false);
      this.set('canLoadMore', !isEmpty(records));
    });
  },
```

**Comment by buschtoens** [comment]
Interesting. We already had this error in #223...

Which versions of ember-light-table, Ember.js and ember-data are you using?

**Comment by trumb1mj** [comment]
[Here is a better description](https://github.com/emberjs/data/issues/3886) of what's going on--this definitely looks like a ember-data issue.

ember: 2.7.0
ember-data: 2.7
ember-light-table: latest

It should be noted that my solution above is a hack and it will eventually break once ember-data removes it from the project.  You should not private functions (_).

**Comment by buschtoens** [comment]
Thanks a lot for referencing that issue! Makes sense, now that I've read https://github.com/emberjs/data/issues/3886#issuecomment-168923038. But I wonder, why this works in the dummy app, since we do work on the [original `RecordArray`](https://github.com/emberjs/data/issues/3886#issuecomment-168923038) as well. 🤔

We should definitely also use the `toArray()` workaround in our dummy app.

**Comment by trumb1mj** [comment]
Where do I find the dummy app code?

**Comment by buschtoens** [comment]
`tests/dummy` or just click the link in my comment :wink:

**Comment by buschtoens** [comment]
@offirgolan Can you please review the [addendum to my original comment](https://github.com/offirgolan/ember-light-table/issues/224#issue-182646785)? :wink:

**Comment by offirgolan** [comment]
@buschtoens 

> Don't use enableSync: true in all examples, but a special one

Lets go with this one. Im not a fan of using private variables such as `_model` in an example.

**Comment by buschtoens** [comment]
I'll send a PR tomorrow night. ✌️

**Comment by jamesdixon** [comment]
Just wanted to mention that I'm also see the issue @trumb1mj mentioned above. I'm using Ember 2.9.1, Ember Data 2.9.0 and ELT 1.6.1. Using the fix @trumb1mj provided resolved the issue.

**Comment by jamesdixon** [comment]
Sorry, forgot to mention I'm using the latest code from the dummy app which uses `toArray()`.

```js
  fetchRecords() {
    this.set('isLoading', true);
    this.get('store').query(this.get('userType'), this.getProperties(['page', 'sort'])).then((records) => {
      this.get('model').pushObjects(records.toArray());
      this.set('canLoadMore', !isEmpty(records));
    }).finally(() => {
      this.set('isLoading', false);
    });
  }
```

**Comment by jamesdixon** [comment]
Hi again 😄  

So, it appears the code from the demo works after all. The change that I made to resolve the issue was something taken from the actual dummy app. In my route, I was originally doing an initial load of the data via:

```js
// route.js
model() {
  return this.store.query(...);
}
```
Once I replaced that call with:

```js
model() {
  return Ember.A([]);
}
```
as done in the dummy app, everything started working properly.

Not sure if this is worth mentioning or showing in the docs, but it might be helpful for those that are trying to do their data loading in the route.
